cpp/src/arrow/engine/substrait/serde_test.cc

```diff
@@ -121,6 +155,57 @@ inline compute::Expression UseBoringRefs(const compute::Expression& expr) {
   return compute::Expression{std::move(modified_call)};
 }
 
+// TODO: complete this interface
+struct TempDataGenerator {
+  TempDataGenerator(const std::shared_ptr<Table> input_table,
+                    const std::string& file_prefix,
+                    std::unique_ptr<arrow::internal::TemporaryDir>& tempdir)
+      : input_table(input_table), file_prefix(file_prefix), tempdir(tempdir) {}
+
+  Status operator()() {
+    auto format = std::make_shared<arrow::dataset::ParquetFileFormat>();
+    auto filesystem = std::make_shared<fs::LocalFileSystem>();
+    const std::string file_name = file_prefix + ".parquet";
+    ARROW_ASSIGN_OR_RAISE(auto file_path, tempdir->path().Join(file_name));
+    data_file_path = file_path.ToString();
+    ARROW_EXPECT_OK(WriteParquetData(data_file_path, filesystem, input_table));
+    return Status::OK();
+  }
+
+  std::shared_ptr<Table> input_table;
+  std::string file_prefix;
+  std::unique_ptr<arrow::internal::TemporaryDir>& tempdir;
+  std::string data_file_path;
+};
+
+void CheckRoundTripResult(const std::shared_ptr<Schema> output_schema,
+                          const std::shared_ptr<Table> expected_table,
+                          compute::ExecContext& exec_context,
+                          std::shared_ptr<Buffer>& buf,
+                          const std::vector<int>& include_columns = {},
+                          const ConversionOptions& conversion_options = {}) {
+  std::shared_ptr<ExtensionIdRegistry> sp_ext_id_reg = MakeExtensionIdRegistry();
+  ExtensionIdRegistry* ext_id_reg = sp_ext_id_reg.get();
+  ExtensionSet ext_set(ext_id_reg);
+  ASSERT_OK_AND_ASSIGN(auto sink_decls, DeserializePlans(
+                                            *buf, [] { return kNullConsumer; },
+                                            ext_id_reg, &ext_set, conversion_options));
+  auto other_declrs = sink_decls[0].inputs[0].get<compute::Declaration>();
+  arrow::AsyncGenerator<util::optional<compute::ExecBatch>> sink_gen;
+  auto sink_node_options = compute::SinkNodeOptions{&sink_gen};
+  auto sink_declaration = compute::Declaration({"sink", sink_node_options, "e"});
+  auto declarations = compute::Declaration::Sequence({*other_declrs, sink_declaration});
+  ASSERT_OK_AND_ASSIGN(auto acero_plan, compute::ExecPlan::Make(&exec_context));
```

Comment:
You should move these lines into `GetTableFromPlan` (maybe change it to `GetTableFromDeclaration`)